students/ThanhNhan/lesson07/html_render.py

```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+import pytest
+"""
+A class-based system for rendering html.
+"""
+
+
+# This is the framework for the base class
+class Element:
+    tag = "html"
+    indent = "    "
+    def __init__(self, content=None, **kwargs):
+       
+        if content is None:
+            self.contents = []
+        else:
+            self.contents = [content]
+        self.attributes = kwargs
+        print("contents is", self.contents)
+        print("kwargs ", self.attributes)
+
+
+    def append(self, new_content):
+        self.contents.append(new_content)
+
+    def render(self, out_file, cur_ind=""):   
+       # self.indent = cur_ind
+        out_file.write(cur_ind + self._open_tag())        
+       # out_file.write("<{}>\n".format(self.tag))
+        for line in self.contents:
+            try:
+                line.render(out_file, cur_ind + self.indent)
+            except AttributeError:
+                out_file.write(cur_ind + self.indent + line)
+                out_file.write("\n")
+        out_file.write(cur_ind + self._close_tag())
+
+    def _open_tag(self):
+        open_tag = []
+        if (self.attributes is None):
+            open_tag.append("<{}>".format(self.tag))
+        else:
+            open_tag.append("<{}".format(self.tag))
+            for key, value in self.attributes.items():       
+                open_tag.append(' {}="{}"'.format(key,value))
+            open_tag.append(">\n")
+        
+        return "".join(open_tag)
+
+    def _close_tag(self):
+        return "</{}>\n".format(self.tag)
+        # out_file.write("just something as a place holder...")
+
+class Html(Element):
+    tag = "html"
+    
+    def render(self, out_file, cur_ind=""):
+        out_file.write(cur_ind + "<!DOCTYPE html>\n")
+        super().render(out_file, cur_ind=cur_ind)
+
+
+class Body(Element):
+    tag = "body"
+
+class P(Element):
+    tag = "p"
+
+class Head(Element):
+    tag = "head"
+
+class Ul(Element):
+    tag = "ul"
+
+class Li(Element):
+    tag = "li"
+
+class OneLineTag(Element):
+
+    def render(self, out_file, cur_ind=""):
+
+        if self.attributes is None:
+            out_file.write(cur_ind + "<{}>".format(self.tag))
+            out_file.write(self.contents[0])
+            out_file.write("</{}>\n".format(self.tag))
+        else:
```

Comment:
there is an _open_tag method in the Element class that you can use here: DRY

(don't repeat yourself).